### PR TITLE
Refactoring the API for multiple interpreter instance support

### PR DESCRIPTION
**Pull request by Gnimuc**
I feel like we also need to split the implementation and use an internal shared header for interfaces.

### CODE REVIEW

File: lib/Interpreter/CppInterOp.cpp [hunk 1]
```diff
@@ -3262,15 +3279,24 @@ namespace Cpp {
     return nullptr;
   }
 
-  void Destruct(TCppObject_t This, TCppScope_t scope, bool withFree /*=true*/) {
-    Decl* Class = (Decl*)scope;
-    if (auto wrapper = make_dtor_wrapper(getInterp(), Class)) {
+  TCppObject_t Construct(TCppScope_t scope, void* arena /*=nullptr*/) {
```
Comment (Gnimuc) [inline lib/Interpreter/CppInterOp.cpp:3282]: also removed the `Impl` suffix.

File: include/clang/Interpreter/CppInterOp.h [hunk 2]
```diff
@@ -78,6 +78,8 @@ namespace Cpp {
   class JitCall {
   public:
     friend CPPINTEROP_API JitCall MakeFunctionCallable(TCppConstFunction_t);
```
Comment (vgvassilev) [inline include/clang/Interpreter/CppInterOp.h:80]: Do we need this as a friend as it is now just forwarding?

File: lib/Interpreter/CppInterOp.cpp [hunk 3]
```diff
@@ -1453,6 +1456,10 @@ namespace Cpp {
     }
   }
 
+  QualType GetBuiltinType(const std::string& name, Sema& sema) {
```
Comment (vgvassilev) [inline lib/Interpreter/CppInterOp.cpp:1459]: This one seems unused and not part of the header. Maybe an oversight from the bigger PR?

File: lib/Interpreter/CppInterOp.cpp [hunk 4]
```diff
@@ -2574,26 +2581,31 @@ namespace Cpp {
     } // namespace
       // End of JitCall Helper Functions
 
-  CPPINTEROP_API JitCall MakeFunctionCallable(TCppConstFunction_t func) {
-    auto* D = (const clang::Decl*)func;
-    if (!D)
-      return {};
+    JitCall MakeFunctionCallableImpl(TInterp_t I, TCppConstFunction_t func) {
+      auto* D = (const clang::Decl*)func;
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply lib/Interpreter/CppInterOp.cpp:2586]: I think that's a good suggestion..

**Review by vgvassilev — APPROVED** [review]
Lgtm! Thank you, @Gnimuc!

**Comment by vgvassilev** [comment]
Can you squash the commits to have clear history?

**Comment by Gnimuc** [comment]
> Can you squash the commits to have clear history?

Done!

**Pull request merged**